File: python/IRB5500.py
from Robot import Robot
import numpy as np
from numpy import pi, sin, cos, arctan2, sqrt


class IRB5500(Robot):

    def __init__(self):
        super().__init__()
        self.dh = [[0, 340, 0, 0], [-pi / 2, 0, 0, -pi / 2], [0, 0, 1300, 0], [-pi / 2, 1507.5, 0, -pi / 2],
                   [0, 79.35, 0, 35 * pi / 180],
                   [0, 79.35, 0, -70 * pi / 180], [-pi / 2, 82.501, 0, 35 * pi / 180]]

        # self.dh = [[0, 340, 0, 0], [-pi / 2, 0, 0, -pi / 2], [0, 0, 1300, 0], [-pi / 2, 1507.5, 0, -pi / 2],
        #            [0, 0, 0, 35 * pi / 180],
        #            [0, 0, 0, -70 * pi / 180], [-pi / 2,0, 0, 35 * pi / 180]]

        self.offset[2, 3] = -212.5

    def cal_t1(self, Td):
        px = Td[0, 3]
        py = Td[1, 3]
        t1 = arctan2(py, px)
        t11 = arctan2(-py, -px)
        res = [t1, t11]
        return res

    def cal_t2(self, Td, t1):
        px = Td[0, 3]
        py = Td[1, 3]
        pz = Td[2, 3]
        a1 = self.dh[1][2]
        d1 = self.dh[0][1]
        T1 = px / cos(t1) - a1
        T3 = pz - d1
        d4 = self.dh[3][1]
        a2 = self.dh[2][2]
        a3 = self.dh[3][2]
        M = (d4 * d4 + a3 * a3 - T1 * T1 - T3 * T3 - a2 * a2) / (2 * a2)
        d = T1 * T1 + T3 * T3 - M * M
        res = []
        if d >= 0:
            t2 = arctan2(T3, -T1) - arctan2(sqrt(d), M)
            t21 = arctan2(T3, -T1) + arctan2(sqrt(d), M)
            res.append(self.normal_range(t2))
            res.append(self.normal_range(t21))
            # res.append(t2 - 2 * pi)
            # res.append(t21 - 2 * pi)
        return res

    def cal_t3(self, Td, t1, t2):
        px = Td[0, 3]
        py = Td[1, 3]
        pz = Td[2, 3]
        a1 = self.dh[1][2]
        d1 = self.dh[0][1]
        d4 = self.dh[3][1]
        a2 = self.dh[2][2]
        a3 = self.dh[3][2]
        T1 = px / cos(t1) - a1
        T3 = pz - d1
        A = T1 - a2 * cos(t2)
        B = -T3 - a2 * sin(t2)
        res = []
        t3 = arctan2(a3 * B - d4 * A, a3 * A + d4 * B) - t2
        res.append(self.normal_range(t3))
        return res

    def cal_t5(self, Td, t1, t2, t3):
        beta = 35 * pi / 180
        ax = Td[0, 2]
        ay = Td[1, 2]
        az = Td[2, 2]
        A = 2 * pow(sin(beta), 4)
        B = 4 * sin(beta) * sin(beta) - 4 * pow(sin(beta), 4)
        C = ax * sin(t2 + t3) * cos(t1) + ay * sin(t2 + t3) * sin(t1) + az * cos(
            t2 + t3) + 2 * pow(sin(beta), 4) - 4 * sin(beta) * sin(beta) + 1
        delta = B * B - 4 * A * C
        res = []
        if delta >= 0:
            D1 = (-B + sqrt(delta)) / (2 * A)
            D2 = (-B - sqrt(delta)) / (2 * A)
            if -1 <= D1 <= 1:
                res.append(arctan2(sqrt(1 - D1 * D1), D1))
                res.append(arctan2(-sqrt(1 - D1 * D1), D1))
            if -1 <= D2 <= 1:
                res.append(arctan2(sqrt(1 - D2 * D2), D2))
                res.append(arctan2(-sqrt(1 - D2 * D2), D2))
        return res

    def cal_t4(self, Td, t1, t2, t3, t5):
        beta = 35 * pi / 180
        ax = Td[0, 2]
        ay = Td[1, 2]
        az = Td[2, 2]
        cb2 = cos(beta) * cos(beta)
        a = -2 * cos(beta) * sin(beta) * (cos(t5) - 1) * (cos(t5) + cb2 - cb2 * cos(t5))
        b = -2 * sin(beta) * sin(t5) * (cos(t5) + cb2 - cb2 * cos(t5))
        c = ay * cos(t1) - ax * sin(t1)
        d = ax * cos(t1) * cos(t2 + t3) + ay * sin(t1) * cos(t2 + t3) - az * sin(t2 + t3)
        t4 = arctan2(a * d - b * c, a * c + b * d)
        res = []
        res.append(t4 + 2 * pi)
        #res.append(t4 + pi)
        res.append(t4)
        #res.append(t4 - pi)
        res.append(t4 - 2 * pi)
        return res

    def cal_t6(self, Td, t1, t2, t3, t5):
        beta = 35 * pi / 180
        nx = Td[0, 0]
        ny = Td[1, 0]
        nz = Td[2, 0]
        ox = Td[0, 1]
        oy = Td[1, 1]
        oz = Td[2, 1]
        cb2 = cos(beta) * cos(beta)
        a = -2 * cos(beta) * sin(beta) * (cos(t5) - 1) * (cos(t5) + cb2 - cb2 * cos(t5))
        b = 2 * sin(beta) * sin(t5) * (cos(t5) + cb2 - cb2 * cos(t5))
        c = -ox * cos(t1) * sin(t2 + t3) - oy * sin(t1) * sin(t2 + t3) - oz * cos(t2 + t3)
        d = -nx * cos(t1) * sin(t2 + t3) - ny * sin(t1) * sin(t2 + t3) - nz * cos(t2 + t3)
        t6 = arctan2(a * d - b * c, a * c + b * d)
        res = []
        res.append(t6 + 2 * pi)
        #res.append(t6 + pi)
        res.append(t6)
        #res.append(t6 - pi)
        res.append(t6 - 2 * pi)
        return res

    def c_to_p(self, thetas):
        thetas[1] = thetas[1] - pi / 2
        thetas[3] = thetas[3] - pi / 2
        thetas[5] = thetas[5] - pi / 2

    def p_to_c(self, thetas):
        thetas[1] = thetas[1] + pi / 2
        thetas[3] = thetas[3] + pi / 2
        thetas[5] = thetas[5] + pi / 2

# if __name__ == '__main__':
#     robot=IRB5500()
#     thetas=np.array([-9.75,-23.51,43.54,8.92,-13.14,-81.58])*pi/180
#     thetas = np.array([0, 0, 55, -8.42, 0, -81.58]) * pi / 180
#     robot.c_to_p(thetas)
#     # T=np.array([[0.0989050558040732,-0.696639309209164,0.710674732916103,1681.65417687322],[-0.773316395688671,0.395522770146976,0.495489998075703,-2106.66221286982],
#     #             [-0.626172917072602,-0.598624940525032,-0.499591670251449,1059.59716586230],[0,0,0,1]])
#     # T=np.array([[-0.0224,-0.3307,0.9435,1085.18],[-0.9310,0.3509,0.1009,-147.53],[-0.3644,-0.8761,-0.3158,946.63],[0,0,0,1]])
#     # s=robot.inv_solution(T,thetas)
#     # robot.p_to_c(s)
#     # for i in s:
#     #     print(i*180/pi,end=" ")
#     # print()
#     T=robot.FK(thetas)
#     print(T)

File: fangzhen/irb5500.cpp
#include "irb5500.h"

IRB5500::IRB5500() {
	//DH参数
	//MDH [θ d a α]
	double dh_v[7][4] = { { 0, 340, 0, 0 },{ -pi / 2, 0, 0, -pi / 2 },{ 0, 0, 1300, 0 },{ -pi / 2, 1507.5, 0, -pi / 2 },
	{ 0,79.35, 0, 35 * pi / 180 },{ 0, 79.35, 0, -70 * pi / 180 },{ -pi / 2,82.501, 0, 35 * pi / 180 } };
	
	for (int i = 0; i < 7; i++) {
		vector<double> dh_i;
		for (int j = 0; j < 4; j++)
			dh_i.push_back(dh_v[i][j]);
		this->dh.push_back(dh_i);
	}
	
	//关节限位
	double joint_lim_v[6][2] = { { -100,100 },{ -30,140 },{ -75,255 },{ -450,450 },{ -360,360 },{ -450,450 } };
	for (int i = 0; i < 6; ++i)
		for (int j = 0; j < 2; ++j)
			joint_lim[i][j] = joint_lim_v[i][j]*pi/180;
	//关节速度限制
	double joint_v_lim_v[6] = { 70.76337004, 62.44863302, 86.84500152, 218.52151492, 53.57735566, 318.63145528 };
	//关节加速度限制
	double joint_a_lim_v[6] = { 536.10351934, 512.16932401, 653.3473183, 3297.1018813, 400.50877485, 2478.01871841 };
	for (int i = 0; i < 6; i++) {
		joint_v_lim[i] = joint_v_lim_v[i];
		joint_a_lim[i] = joint_a_lim_v[i];
	}

	MatrixXd offset_v(4, 4);
	offset_v.setIdentity();
	offset_distance = -212.5; //偏移距离
	offset_v(2, 3) = offset_distance; 
	offset = offset_v; //偏移距离矩阵
	robot_type = "ABB";
}


vector<double> IRB5500::cal_t5(MatrixXd Td, double theta1, double theta2, double theta3) {
	double beta = 35 * pi / 180;
	double ax = Td(0, 2);
	double ay = Td(1, 2);
	double az = Td(2, 2);
	double A = 2 * pow(sin(beta), 4);
	double B = 4 * sin(beta)*sin(beta) - 4 * pow(sin(beta), 4);
	double C = ax*sin(theta2 + theta3)*cos(theta1) + ay*sin(theta2 + theta3)*sin(theta1) + az*cos(theta2 + theta3) + 2 * pow(sin(beta), 4) - 4 * sin(beta)*sin(beta) + 1;
	double delta = B*B - 4 * A*C;
	vector<double> res;
	if (delta >= 0) {
		double D1 = (-B + sqrt(delta)) / (2 * A);
		double D2 = (-B - sqrt(delta)) / (2 * A);
		if (-1 <= D1 && D1 <= 1) {
			res.push_back(atan2(sqrt(1 - D1*D1), D1));
			res.push_back(atan2(-sqrt(1 - D1*D1), D1));
		}
		if (-1 <= D2 && D2 <= 1) {
			res.push_back(atan2(sqrt(1 - D2*D2), D2));
			res.push_back(atan2(-sqrt(1 - D2*D2), D2));
		}
	}
	return res;
}

vector<double> IRB5500::cal_t4(MatrixXd Td, double theta1, double theta2, double theta3, double theta5) {
	double beta = 35 * pi / 180;
	double ax = Td(0, 2);
	double ay = Td(1, 2);
	double az = Td(2, 2);
	double cb2 = cos(beta)*cos(beta);
	double a = -2 * cos(beta)*sin(beta)*(cos(theta5) - 1)*(cos(theta5) + cb2 - cb2*cos(theta5));
	double b = -2 * sin(beta)*sin(theta5)*(cos(theta5) + cb2 - cb2*cos(theta5));
	double c = ay*cos(theta1) - ax*sin(theta1);
	double d = ax*cos(theta1)*cos(theta2 + theta3) + ay*sin(theta1)*cos(theta2 + theta3) - az*sin(theta2 + theta3);
	double t4 = atan2(a*d - b*c, a*c + b*d);
	vector<double> res;
	res.push_back(t4);
	res.push_back(t4 - 2 * pi);
	if(t4<=0)
		res.push_back(t4 + 2 * pi);
	return res;
}

vector<double> IRB5500::cal_t6(MatrixXd Td, double theta1, double theta2, double theta3, double theta5) {
	double beta = 35 * pi / 180;
	double nx = Td(0, 0);
	double ny = Td(1, 0);
	double nz = Td(2, 0);
	double ox = Td(0, 1);
	double oy = Td(1, 1);
	double oz = Td(2, 1);

	double cb2 = cos(beta)*cos(beta);
	double a = -2 * cos(beta)*sin(beta)*(cos(theta5) - 1)*(cos(theta5) + cb2 - cb2*cos(theta5));
	double b = 2 * sin(beta)*sin(theta5)*(cos(theta5) + cb2 - cb2*cos(theta5));
	double c = -ox*cos(theta1)*sin(theta2 + theta3) - oy*sin(theta1)*sin(theta2 + theta3) - oz*cos(theta2 + theta3);
	double d = -nx*cos(theta1)*sin(theta2 + theta3) - ny*sin(theta1)*sin(theta2 + theta3) - nz*cos(theta2 + theta3);
	double t6 = atan2(a*d - b*c, a*c + b*d);
	vector<double> res;
	res.push_back(t6);
	res.push_back(t6 - 2 * pi);
	if(t6<=0)
		res.push_back(t6 + 2 * pi);
	return res;
}

//控制器角度值转为程序中的角度值
void IRB5500::c_to_p(double* thetas) {
	thetas[1] = thetas[1] - pi / 2;
	thetas[3] = thetas[3] - pi / 2;
	thetas[5] = thetas[5] - pi / 2;
}

//程序中的角度值转为控制器中的角度值
void IRB5500::p_to_c(double* thetas) {
	thetas[1] = thetas[1] + pi / 2;
	thetas[3] = thetas[3] + pi / 2;
	thetas[5] = thetas[5] + pi / 2;
}

//程序角度值转为控制器角度值(角度单位的)
void IRB5500::p_to_c_deg(double* thetas) {
	thetas[1] = thetas[1] + 90;
	thetas[3] = thetas[3] + 90;
	thetas[5] = thetas[5] + 90;
}